swarm/src/behaviour/peer_addresses.rs

```diff
@@ -6,16 +6,58 @@ use libp2p_identity::PeerId;
 
 use crate::{behaviour::FromSwarm, DialError, DialFailure, NewExternalAddrOfPeer};
 
+#[derive(Debug, Clone)]
+/// Configuration of a [`PeerAddresses`] instance.
+pub struct PeerAddressesConfig {
+    /// Capacity of the [`PeerAddresses`] cache.
+    number_of_peers: NonZeroUsize,
+
+    /// Maximum number of cached addresses per peer.
+    number_of_addresses_per_peer: NonZeroUsize,
+}
+
+impl PeerAddressesConfig {
+    /// Configure the capacity of the [`PeerAddresses`] cache.
+    /// The default capacity is 100.
+    pub fn with_number_of_peers(mut self, number_of_peers: NonZeroUsize) -> Self {
+        self.number_of_peers = number_of_peers;
+        self
+    }
+
+    /// Configure the maximum number of cached addresses per peer.
+    /// The default number is 10.
+    pub fn with_number_of_addresses_per_peer(
+        mut self,
+        number_of_addresses_per_peer: NonZeroUsize,
+    ) -> Self {
+        self.number_of_addresses_per_peer = number_of_addresses_per_peer;
+        self
+    }
+}
+
+impl Default for PeerAddressesConfig {
+    fn default() -> Self {
+        Self {
+            number_of_peers: NonZeroUsize::new(100).expect("100 != 0"),
+            number_of_addresses_per_peer: NonZeroUsize::new(10).expect("10 != 0"),
+        }
+    }
+}
+
 /// Struct for tracking peers' external addresses of the [`Swarm`](crate::Swarm).
 #[derive(Debug)]
-pub struct PeerAddresses(LruCache<PeerId, LruCache<Multiaddr, ()>>);
+pub struct PeerAddresses {
+    config: PeerAddressesConfig,
+    inner: LruCache<PeerId, LruCache<Multiaddr, ()>>,
+}
 
 impl PeerAddresses {
     /// Creates a [`PeerAddresses`] cache with capacity for the given number of peers.
     ///
-    /// For each peer, we will at most store 10 addresses.
-    pub fn new(number_of_peers: NonZeroUsize) -> Self {
-        Self(LruCache::new(number_of_peers.get()))
+    /// For each peer, we will at most store `config.number_of_addresses_per_peer` addresses.
+    pub fn new(config: PeerAddressesConfig) -> Self {
+        let inner = LruCache::new(config.number_of_peers.get());
+        Self { config, inner }
```

Comment:
Ah sorry, I just noticed now that this is a breaking change. 
Breaking changes in the swarm are quite painful, because we need to do a minor bump and release for every single behavior protocol that depends on the swarm.

We could avoid the breaking change by not adding the `PeerAddressesConfig` here and instead just add `with_number_of_addresses_per_peer` directly for `PeerAddresses`.
The `PeerAddressesConfig` could still live as `CacheConfig` in `identity`.

Wdyt? I understand that by now you might not be working on this anymore, so I can also push this change if folks agree with it. cc @jxs